### PR TITLE
feat: Drawer add loading prop to show spinner

### PR DESCRIPTION
**Pull request by Enigama**
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [x] Site / documentation update
- [x] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [x] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 💡 Background and solution
It would be nice to have loading prop, it will solve some issue when async data loads in drawer and we don't want to show drawer without title or some body parts.

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by Enigama** [comment]
@yoyo837 @afc163

File: components/drawer/DrawerPanel.tsx [hunk 1]
```diff
@@ -87,6 +106,19 @@ const DrawerPanel: React.FC<DrawerPanelProps> = (props) => {
     },
   );
 
+  // >>>>>>>>> Spinning
+  let spinProps: SpinProps | undefined;
+  if (typeof loading === 'boolean') {
+    spinProps = {
+      spinning: loading,
+    };
+  } else if (typeof loading === 'object') {
```
Comment (yoyo837) [inline components/drawer/DrawerPanel.tsx:115]: Need test case to cover this scene.

File: components/drawer/__tests__/Drawer.test.tsx [hunk 2]
```diff
@@ -179,6 +179,29 @@ describe('Drawer', () => {
...
+    });
+    it('have a spinner with custom text', () => {
+      const { container: wrapper } = render(
+        <Drawer open loading={{ tip: 'Loading...' }} getContainer={false}>
+          Here is content of Drawer
+        </Drawer>,
+      );
+
+      triggerMotion();
+      expect(wrapper.firstChild).toMatchSnapshot();
```
Comment (yoyo837) [inline components/drawer/__tests__/Drawer.test.tsx:201]: `expect` the `tip` value in DOM should be `Loading...`.

**Comment by Enigama** [comment]
@yoyo837 @afc163 I think over comment that we should prevent to show drawer content while `loading` here is my thought how to do it, maybe you have some suggestion?
`if (spinProps?.spinning) {
    return (
      <Spin
        spinning={false}
        style={{
          height: '100%',
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
        }}
        {...spinProps}
      />
    );
  }

  return (
    <>
      {headerNode}
      <div
        className={classNames(
          `${prefixCls}-body`,
          drawerClassNames?.body,
          drawerContext?.classNames?.body,
        )}
        style={{
          ...drawerContext?.styles?.body,
          ...bodyStyle,
          ...drawerStyles?.body,
        }}
      >
        {children}
      </div>
      {footerNode}
    </>
  );`

**Comment by Enigama** [comment]
@yoyo837 @afc163 @zombieJ what your thoughts?

**Comment by Enigama** [comment]
Is it ok that regression and tests are failed?

File: components/drawer/__tests__/Drawer.test.tsx [hunk 3]
```diff
@@ -179,6 +179,30 @@ describe('Drawer', () => {
...
+    it('have a spinner with custom text', () => {
+      const tip = 'Loading...';
+      const { getByText } = render(
+        <Drawer open loading={{ tip }} getContainer={false}>
+          Here is content of Drawer
+        </Drawer>,
+      );
+
+      triggerMotion();
+      expect(getByText(tip)).toBeInTheDocument();
```
[the remark this answers is not in the document]
Comment (Enigama) [reply components/drawer/__tests__/Drawer.test.tsx:202]: If we agree that we should block content with spinner than we will not have tip property and will not have this test(maybe will be another one)

**Comment by afc163** [comment]
CI failed: https://github.com/ant-design/ant-design/actions/runs/8781495679/job/24093945327?pr=48563

**Comment by Enigama** [comment]
@yoyo837 @afc163 I see that ci failed on `demo-extend.test.tsx` which use original `rc-drawer` where we do not have loading statement and there fore snapshot has differences, what should i do, add conditional rendering in `rc-drawer` as well as we have in ant-drawer?
Like:
image

**Comment by Enigama** [comment]
@yoyo837 @afc163 Should I add loading to original component `rc-drawer` because I think it would fix test?

**Comment by Enigama** [comment]
:sob:

**Comment by Enigama** [comment]
I've done the command: `npm run test:site-update` which was suggested by the failed e2e test in gh action, did I something wrong?

**Comment by Enigama** [comment]
@MadCcc @afc163 @yoyo837 @zombieJ what should I do?

**Comment by Enigama** [comment]
@MadCcc @afc163 @yoyo837 @zombieJ  I see that check `test / test (18, dom, 1/2) (pull_request)` has failed but it doesn't concern to my changes(I guess), can you guys check the RP and if it's fine than approve it, if it's not, please give me detailed message it would help me to fix it a way more faster. Thanks.

**Comment by yoyo837** [comment]
`React` recently released `18.3.0`, Thye `CI` needs to be adjusted, please wait https://github.com/ant-design/ant-design/pull/48640

**Comment by Enigama** [comment]
[waiting-pr](https://github.com/ant-design/ant-design/pull/48661)

**Comment by Enigama** [comment]
@yoyo837 Is solution ok?

File: components/drawer/index.zh-CN.md [hunk 4]
```diff
@@ -45,7 +46,7 @@ v5 使用 `rootClassName` 与 `rootStyle` 来配置最外层元素样式。原 v
 :::
 
 | 参数 | 说明 | 类型 | 默认值 | 版本 |
-| --- | --- | --- | --- | --- |
+| --- | --- | --- | --- | --- | --- |
```
Comment (yoyo837) [inline components/drawer/index.zh-CN.md:49]: Redundant column.

File: components/drawer/index.zh-CN.md [hunk 5]
```diff
@@ -67,6 +68,7 @@ v5 使用 `rootClassName` 与 `rootStyle` 来配置最外层元素样式。原 v
 | style | 设计 Drawer 容器样式，如果你只需要设置内容部分请使用 `bodyStyle` | CSSProperties | - |  |
 | styles | 语义化结构 style | [Record<SemanticDOM, CSSProperties>](#semantic-dom) | - | 5.10.0 |
 | title | 标题 | ReactNode | - |  |
+| loading | 显示旋转指示器 | boolean \| `Omit<SpinProp, 'fullScreen' | 'tip'>` | false | 5.2.0 |
```
Comment (yoyo837) [inline components/drawer/index.zh-CN.md:71]:
```suggestion
| loading | 显示旋转指示器 | boolean \| `Omit<SpinProp, 'fullScreen' | 'tip'>` | false | 5.17.0 |
```

**Comment by Enigama** [comment]
@yoyo837 I see that e2e has failed, I locally ran the command to update snap `test:site-update` it's again in `scripts` folder, should i commit it? Or it's fine?
image

**Review by afc163 — APPROVED** [review]

**Pull request merged**